sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SerializationDiagnosticsContext.java

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation;
+
+import com.azure.cosmos.ZonedDateTimeSerializer;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.ser.std.StdSerializer;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public class SerializationDiagnosticsContext {
+    public volatile List<SerializationDiagnosticsContext.SerializationDiagnostics> serializationDiagnosticsList;
+
+    public void addMetaDataDiagnostic(SerializationDiagnosticsContext.SerializationDiagnostics serializationDiagnostics) {
+        if (serializationDiagnosticsList == null) {
+            serializationDiagnosticsList = new ArrayList<>();
+        }
+
+        serializationDiagnosticsList.add(serializationDiagnostics);
+    }
+
+    public <T> T getResource(Callable<T> function, SerializationType serializationType) {
+        ZonedDateTime serializationStartTime = ZonedDateTime.now(ZoneOffset.UTC);
+        T t = null;
+        try {
+            t = function.call();
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
```

Comment:
as you are invoking this in the deserialization codepath it should have the same behaviour in terms of exception throwing.

This changes the exception type.

Reply:
All the places where this method is getting called are eventually throwing IllegalArgumentException via  JsonSerializable.  So either we throw  IllegalArgumentException  or let it be the way we have now IllegalStateException wraps IllegalArgumentException, thoughts ? And if we agree on removing getResource() method altogether measuring pros and cons, that this problem will be solved itself.